crates/rust-analyzer/src/cli/diagnostics.rs

```diff
@@ -0,0 +1,70 @@
+//! Analyze all modules in a project for diagnostics. Exits with a non-zero status
+//! code if any errors are found.
+
+use anyhow::anyhow;
+use ra_db::{SourceDatabase, SourceDatabaseExt};
+use ra_ide::Severity;
+use std::{collections::HashSet, path::Path};
+
+use crate::cli::{load_cargo::load_cargo, Result};
+use hir::Semantics;
+
+pub fn diagnostics(path: &Path, load_output_dirs: bool, all: bool) -> Result<()> {
+    let (host, roots) = load_cargo(path, load_output_dirs)?;
+    let db = host.raw_database();
+    let analysis = host.analysis();
+    let semantics = Semantics::new(db);
+    let members = roots
+        .into_iter()
+        .filter_map(|(source_root_id, project_root)| {
+            // filter out dependencies
+            if project_root.is_member() {
+                Some(source_root_id)
+            } else {
+                None
+            }
+        })
+        .collect::<HashSet<_>>();
+
+    let mut found_error = false;
+    let mut visited_files = HashSet::new();
+    let crate_graph = db.crate_graph();
+    for crate_id in crate_graph.iter() {
+        let krate = &crate_graph[crate_id];
+        if let Some(crate_name) = &krate.display_name {
+            println!("processing crate: {}", crate_name);
+        } else {
+            println!("processing crate: unknown");
+        }
+        for file_id in db.source_root(db.file_source_root(krate.root_file_id)).walk() {
```

Comment:
a source root can contain multiple crates, and I think you'll walk all files in the source root for each crate here. Why not just go through `members` in the first place? you can still get the crate of the module from the module you got from `to_module_def`.

Reply:
> you can still get the crate of the module from the module you got from to_module_def

So `to_module_def` gives me a `Module`, which I can use to get a `hir::code_model::Crate`. But to get the crate name I need `CrateData`, and the way I've been getting that is by looking up the `CrateId` in the crate graph. For this reason, I had to modify the visibility of the `id` field of `hir::code_model::Crate` from `pub(crate)` to `pub`. 

Does this seem correct? See my latest commit for the implementation of this.

Reply:
Hmm no we shouldn't make that `pub`... instead we can add an accessor for `display_name` to `Crate` if it doesn't have one yet.

Reply:
I've reverted to `pub(crate)` and added an accessor here as you suggested.